### PR TITLE
sql: add volatility information for operators

### PR DESCRIPTION
**Pull request by RaduBerinde**
This change adds volatility information for unary, comparison, and binary
operators. A subset of comparison operators are leak-proof; a small set of
operators (mostly those involving the current timezone) are stable; and the rest
are immutable.

Release note: None

### CODE REVIEW

File: pkg/sql/sem/tree/eval.go [hunk 1]
```diff
@@ -1852,70 +1986,74 @@ func (o cmpOpOverload) LookupImpl(left, right *types.T) (*CmpOp, bool) {
...
-		makeEqFn(types.Timestamp, types.Timestamp),
-		makeEqFn(types.TimestampTZ, types.TimestampTZ),
-		makeEqFn(types.Uuid, types.Uuid),
-		makeEqFn(types.VarBit, types.VarBit),
+		makeEqFn(types.AnyEnum, types.AnyEnum, VolatilityImmutable),
+		makeEqFn(types.Bool, types.Bool, VolatilityLeakProof),
+		makeEqFn(types.Bytes, types.Bytes, VolatilityLeakProof),
+		makeEqFn(types.Date, types.Date, VolatilityLeakProof),
+		makeEqFn(types.Decimal, types.Decimal, VolatilityImmutable),
+		makeEqFn(types.AnyCollatedString, types.AnyCollatedString, VolatilityLeakProof),
```
Comment (otan) [inline pkg/sql/sem/tree/eval.go:2025]: this one doesn't feel leak proof if the collated strings don't have the same encoding, am i wrong?

Comment (otan) [reply pkg/sql/sem/tree/eval.go:2025]: psql:
```

otan=# select 'a' collate "C" = 'b' collate "fr_FR";
ERROR:  collation mismatch between explicit collations "C" and "fr_FR"
LINE 1: select 'a' collate "C" = 'b' collate "fr_FR";
```

crdb:
```
root@127.0.0.1:63654/movr> select 'a' collate "en_AU" = 'b' collate "fr_FR";
ERROR: unsupported comparison operator: <collatedstring{en_AU}> = <collatedstring{fr_FR}>
SQLSTATE: 22023
```

guess we don't even resolve that comparator at overload resolution so in a way it's leakproof once it reaches that point?

**Review by otan — APPROVED** [review]
LGTM if collated string check looks sane to you.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49764)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/tree/eval.go, line 2025 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49764#-M8l7n2t1r5Vl8OZvMgm-r1-2025:-M8lMmJKEqeZNg3bI28s:bqvav7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/88b938c5c3edb35878177eb19b5c7077aebbf0a6/pkg/sql/sem/tree/eval.go#L2025)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

psql:
```

otan=# select 'a' collate "C" = 'b' collate "fr_FR";
ERROR:  collation mismatch between explicit collations "C" and "fr_FR"
LINE 1: select 'a' collate "C" = 'b' collate "fr_FR";
```

crdb:
```
root@127.0.0.1:63654/movr> select 'a' collate "en_AU" = 'b' collate "fr_FR";
ERROR: unsupported comparison operator: <collatedstring{en_AU}> = <collatedstring{fr_FR}>
SQLSTATE: 22023
```

guess we don't even resolve that comparator at overload resolution so in a way it's leakproof once it reaches that point?
</blockquote></details>

Interesting point. I think they are indeed leak-proof given that type checking will prevent it being used in that way. Added comments to that effect.

**Pull request merged**